// angular8-yandex-maps/esm2015/lib/services/ya-geocoder/ya-geocoder.service.js
import { Injectable, NgZone } from '@angular/core';
import { switchMap } from 'rxjs/operators';
import { from, Observable } from 'rxjs';
import { YaApiLoaderService } from '../ya-api-loader/ya-api-loader.service';
import * as i0 from "@angular/core";
import * as i1 from "../ya-api-loader/ya-api-loader.service";
/**
 * The `YaGeocoder` service wraps `ymaps.geocode` static function from the Yandex Maps API.
 *
 * @example
 * ```ts
 * import { YaGeocoderService } from 'angular8-yandex-maps';
 *
 * export class AppComponent {
 *   constructor(private yaGeocoderService: YaGeocoderService) {
 *     this.yaGeocoderService.geocode('Moscow')
 *       .subscribe(v => console.log(v))
 *   }
 * }
 * ```
 */
export class YaGeocoderService {
    constructor(_ngZone, _yaApiLoaderService) {
        this._ngZone = _ngZone;
        this._yaApiLoaderService = _yaApiLoaderService;
    }
    /**
     * Processes geocoding requests
     * @param request
     * @param options
     */
    geocode(request, options) {
        return this._yaApiLoaderService.load().pipe(switchMap(() => from(ymaps.geocode(request, options))), switchMap((result) => new Observable((observer) => {
            this._ngZone.run(() => {
                observer.next(result);
                observer.complete();
            });
        })));
    }
}
YaGeocoderService.ɵprov = i0.ɵɵdefineInjectable({ factory: function YaGeocoderService_Factory() { return new YaGeocoderService(i0.ɵɵinject(i0.NgZone), i0.ɵɵinject(i1.YaApiLoaderService)); }, token: YaGeocoderService, providedIn: "root" });
YaGeocoderService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
YaGeocoderService.ctorParameters = () => [
    { type: NgZone },
    { type: YaApiLoaderService }
];
